mono/metadata/icall.c

```diff
@@ -248,29 +248,32 @@ array_set_value_impl (MonoArrayHandle arr, MonoObjectHandle value, guint32 pos,
 	guint64 u64 = 0;
 	gint64 i64 = 0;
 	gdouble r64 = 0;
+	gboolean castOk = FALSE;
 
-	uint32_t arr_gchandle = 0;
-	uint32_t value_gchandle = 0;
+	MONO_ENTER_NO_SAFEPOINTS;
```

Comment:
I think this no safepoints region is too big.  until line 269 you don't need it.

There is a cost to "no safepoints" regions - they increase how long the GC has to wait for threads to suspend.  It is better to keep them small.

Additionally, the runtime will assert if try to switch to GC Safe, or invoke managed code, or try to allocate something while in a no safepoints region.